contracts/0.8.9/oracle/AccountingOracle.sol

```diff
@@ -0,0 +1,817 @@
+            revert NumExitedValidatorsCannotDecrease();
+        }
+
+        uint256 exitedValidatorsPerDay =
+            (exitedValidators - prevExitedValidators) * (1 days) /
+            (SECONDS_PER_SLOT * slotsElapsed);
+
+        if (exitedValidatorsPerDay > boundaries.maxExitedValidatorsPerDay) {
+            revert ExitedValidatorsLimitExceeded(
+                boundaries.maxExitedValidatorsPerDay,
+                exitedValidatorsPerDay
+            );
+        }
+
+        stakingRouter.updateExitedKeysCountByStakingModule(
+            stakingModuleIds,
+            numExitedValidatorsByStakingModule
+        );
+    }
+
+    function _submitReportExtraDataList(uint256[] calldata items) internal {
+        _checkMsgSenderIsAllowedToSubmitData();
+        _checkProcessingDeadline();
+
+        ExtraDataProcessingState memory procState = _storageExtraDataProcessingState().value;
+
+        if (procState.refSlot != LAST_PROCESSING_REF_SLOT_POSITION.getStorageUint256()) {
+            revert CannotSubmitExtraDataBeforeMainData();
+        }
+
+        if (procState.itemsProcessed == procState.itemsCount) {
+            revert ExtraDataAlreadyProcessed();
+        }
+
+        if (procState.itemsProcessed != 0) {
+            revert ExtraDataListOnlySupportsSingleTx();
+        }
+
+        if (items.length != procState.itemsCount) {
+            revert UnexpectedExtraDataItemsCount(procState.itemsCount, items.length);
+        }
+
+        if (procState.dataFormat != EXTRA_DATA_FORMAT_LIST) {
+            revert UnexpectedExtraDataFormat(procState.dataFormat, EXTRA_DATA_FORMAT_LIST);
+        }
+
+        bytes32 dataHash = MemUtils.keccakUint256Array(items);
+        if (dataHash != procState.dataHash) {
+            revert UnexpectedDataHash(procState.dataHash, dataHash);
+        }
+
+        ExtraDataProcessingState storage _procState = _storageExtraDataProcessingState().value;
+
+        _procState.lastProcessedItem = _processExtraDataItems(items, 0, 0);
+        _procState.itemsProcessed = uint64(items.length);
+
+        emit ExtraDataSubmitted(procState.refSlot, items.length, items.length);
+    }
+
+    struct ExtraDataIterState {
+        uint256 firstItemIndex;
+        uint256 nextIndex;
+        int256 lastType;
+        int256 lastModuleId;
+        int256 lastNodeOpId;
+        ResizableArray.Array nopIds;
+        ResizableArray.Array keyCounts;
+    }
+
+    function _processExtraDataItems(
+        uint256[] calldata items,
+        uint256 itemsProcessed,
+        uint256 lastProcessedItem
+    ) internal returns (uint256) {
+        IStakingRouter stakingRouter = IStakingRouter(LOCATOR.stakingRouter());
+
+        ExtraDataIterState memory iter = ExtraDataIterState({
+            firstItemIndex: itemsProcessed,
+            nextIndex: 0,
+            lastType: -1,
+            lastModuleId: -1,
+            lastNodeOpId: -1,
+            nopIds: ResizableArray.preallocate(20),
```

Comment:
Do we really need to have `ResizableArray ` here?

Reply:
why not just send this from oracle through the call data?

Reply:
we can do without it if we oblige the oracle daemon to pass the total numbers of validators with exited keys for each module. but we'll then need to allocate a new memory array for each module which seems like a waste of memory and thus gas